### PR TITLE
feat: create symlink for build files present on node modules installed with relative paths

### PR DESCRIPTION
**Pull request by mistic**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?
Currently if a local module installed with npm contains a BUILD file we generate the usual BUILD file content and just append the one found on the BUILD file. 

Issue Number: https://github.com/bazelbuild/rules_nodejs/issues/2131

## What is the new behavior?

With this PR the BUILD file inside an installed local module will be symlinked instead of generated and get its content appended.

## Does this PR introduce a breaking change?

- [x] Yes
- [ ] No

It is really unlikely that the current behaviour is being useful to someone who relies on having a BUILD file in a local installed npm module as it introduces a race condition. However, the migration path would be to rewrite their BUILD files inside such modules to contain the expected generated targets by npm_install/yarn_install rules like `*-module__contents` or `*--module__files` and also add the extra targets they have previously defined on the BUILD file. 

HINT: In order to don't have to manually writ `*-module__contents` or `*--module__files` we can remove the BUILD file, run the npm_install/yarn_install rules and just grasp the generated BUILD file content.

### CODE REVIEW

**Comment by alexeagle** [comment]
Awesome, this looks good to me. Thanks! I think I agree that no one really should rely on the current behavior but @gregmagolan added that and has a better memory than I do.

File: internal/npm_install/generate_build_file.ts [hunk 1]
```diff
@@ -194,11 +203,33 @@ js_library(
...
   if (pkg._files.includes('BUILD.bazel')) buildFilePath = 'BUILD.bazel';
+
+  // Recreate the pkg dir inside the node_modules folder
+  const nodeModulesPkgDir = `node_modules/${pkg._dir}`;
+  // Check if the current package dep dir is a symlink (which happens when we install a node_module
+  // with link:)
+  const isPkgDirASymlink =
+      fs.existsSync(nodeModulesPkgDir) && fs.lstatSync(nodeModulesPkgDir).isSymbolicLink();
+  // Check if the current package is also written inside the workspace
+  const isPkgInsideWorkspace = fs.realpathSync(nodeModulesPkgDir).includes(fs.realpathSync(`.`));
```
Comment (gregmagolan) [inline internal/npm_install/generate_build_file.ts:220]: Its a corner case, but if `symlink_node_modules = False` then yarn & npm are run in the root of the external repository. In that case, `link:` would be fairly useless as anything linked with a relative path would have to be copied over via that `data` attribute. Still, the comment here should reflect that `fs.realpathSync(`.`)` will not be the root of the workspace if `symlink_node_modules = False`

**Comment by gregmagolan** [comment]
This looks great. I read through the issue and the code and agree that this is the right thing to do if the node_module is a symlink to inside the workspace. It's a clever way to consume 1st party deps. The only thing I'd add is to update the yarn_install & npm_install docs to document this feature & the reasoning.

As an aside, why use this method instead of `js_library` with `package_name` to link 1st party deps? End result of a link in node_modules to the 1st party dep seems the same tho the js_library with package_name would use the linker to achieve it.

**Comment by mistic** [comment]
@gregmagolan Thanks for reviewing this as you are one of the few people well positioned and with enough knowledge to validate the idea. Regarding your question, `js_library` with `package_name` could be used to solve almost every situation. The problem arises when a local package consumed by the monorepo also gets published into the npm registry and starts being used by other `node_module` that you also have installed in your monorepo. That will cause a race condition that will make you not being able to use the linked node_module correctly as the generated BUILD file is not updated yet. If you run npm/yarn install a second time it will succeed because the BUILD file on the second run gots updated. Fixing the legacy behaviour of appending BUILD file contents and link instead will empower us to cover all the situations we could end up having in a node environment.

File: internal/npm_install/generate_build_file.ts (hunk 1 above)
Comment (mistic) [reply internal/npm_install/generate_build_file.ts:220]: @gregmagolan have you any suggestion to solve that corner case?

**Comment by mistic** [comment]
@alexeagle I've also added some content both to the `yarn_install` and `npm_install` rules about that new behaviour. Can you take a look and review it please?

**Comment by alexeagle** [comment]
I'm still not sure this use case of bazel -> bazel dependency makes sense, even given your problem description I still think you should have a `js_library` with `package_name` as the dependency of any Bazel code. That should be orthogonal to whether some non-Bazel dependencies in your repo still depend on the npm-built package.

File: docs/Built-ins.md [hunk 2]
```diff
@@ -947,6 +947,24 @@ Defaults to `True`
 
 Defaults to `3600`
 
+**LOCAL MODULES**
```
Comment (alexeagle) [inline docs/Built-ins.md:950]: see the note at the top of this file - it's generated so your edits will be lost. You should instead change the stardoc blocks in the source file

Comment (mistic) [reply docs/Built-ins.md:950]: @alexeagle You're right. Any idea about the best place to put those notes as they are not doc for any attribute?

Comment (mistic) [reply docs/Built-ins.md]: I've updated the docs correctly on the source

File: internal/npm_install/generate_build_file.ts (hunk 1 above)
Comment (alexeagle) [reply internal/npm_install/generate_build_file.ts:224]: Looking for approve from @gregmagolan to merge as-is

**Review by gregmagolan — APPROVED** [review]

**Pull request merged**